rpcs3/Emu/Cell/lv2/sys_lwcond.cpp

```diff
@@ -102,125 +155,149 @@ error_code _sys_lwcond_signal(ppu_thread& ppu, u32 lwcond_id, u32 lwmutex_id, u6
 		fmt::throw_exception("Unknown mode (%d)", mode);
 	}
 
-	const auto cond = idm::check<lv2_obj, lv2_lwcond>(lwcond_id, [&, notify = lv2_obj::notify_all_t()](lv2_lwcond& cond) -> int
+	while (true)
 	{
-		ppu_thread* cpu = nullptr;
-
-		if (ppu_thread_id != u32{umax})
+		if (ppu.test_stopped())
 		{
-			cpu = idm::check_unlocked<named_thread<ppu_thread>>(static_cast<u32>(ppu_thread_id));
-
-			if (!cpu)
-			{
-				return -1;
-			}
+			ppu.state += cpu_flag::again;
+			return {};
 		}
 
-		lv2_lwmutex* mutex = nullptr;
+		bool finished = true;
+
+		ppu.state += cpu_flag::wait;
 
-		if (mode != 2)
+		const auto cond = idm::check<lv2_obj, lv2_lwcond>(lwcond_id, [&, notify = lv2_obj::notify_all_t()](lv2_lwcond& cond) -> int
 		{
-			mutex = idm::check_unlocked<lv2_obj, lv2_lwmutex>(lwmutex_id);
+			ppu_thread* cpu = nullptr;
 
-			if (!mutex)
+			if (ppu_thread_id != u32{umax})
 			{
-				return -1;
+				cpu = idm::check_unlocked<named_thread<ppu_thread>>(static_cast<u32>(ppu_thread_id));
+
+				if (!cpu)
+				{
+					return -1;
+				}
 			}
-		}
 
-		if (atomic_storage<ppu_thread*>::load(cond.sq))
-		{
-			std::lock_guard lock(cond.mutex);
+			lv2_lwmutex* mutex = nullptr;
 
-			if (cpu)
+			if (mode != 2)
 			{
-				if (static_cast<ppu_thread*>(cpu)->state & cpu_flag::again)
+				mutex = idm::check_unlocked<lv2_obj, lv2_lwmutex>(lwmutex_id);
+
+				if (!mutex)
 				{
-					ppu.state += cpu_flag::again;
-					return 0;
+					return -1;
 				}
 			}
 
-			auto result = cpu ? cond.unqueue(cond.sq, cpu) :
-				cond.schedule<ppu_thread>(cond.sq, cond.protocol);
-
-			if (result)
+			//if (atomic_storage<ppu_thread*>::load(cond.sq))
 			{
-				if (static_cast<ppu_thread*>(result)->state & cpu_flag::again)
+				std::lock_guard lock(cond.mutex);
+
+				if (ppu.state & cpu_flag::suspend)
 				{
-					ppu.state += cpu_flag::again;
+					finished = false;
 					return 0;
 				}
 
-				if (mode == 2)
+				if (cpu)
 				{
-					static_cast<ppu_thread*>(result)->gpr[3] = CELL_EBUSY;
+					if (static_cast<ppu_thread*>(cpu)->state & cpu_flag::again)
+					{
+						ppu.state += cpu_flag::again;
+						return 0;
+					}
 				}
 
-				if (mode != 2)
+				auto result = cpu ? cond.unqueue(cond.sq, cpu) :
+					cond.schedule<ppu_thread>(cond.sq, cond.protocol);
+
+				if (result)
 				{
-					if (mode == 3 && mutex->load_sq()) [[unlikely]]
+					if (static_cast<ppu_thread*>(result)->state & cpu_flag::again)
 					{
-						std::lock_guard lock(mutex->mutex);
+						ppu.state += cpu_flag::again;
+						return 0;
+					}
 
-						// Respect ordering of the sleep queue
-						mutex->try_own(result, true);
-						auto result2 = mutex->reown<ppu_thread>();
+					if (mode == 2)
+					{
+						static_cast<ppu_thread*>(result)->gpr[3] = CELL_EBUSY;
+					}
 
-						if (result2->state & cpu_flag::again)
+					if (mode != 2)
```

Comment:
can mode change?
otherwise:
```suggestion
					else
```

Reply:
Fixed.